crates/federated-server/src/server/graph_updater/schema_file.rs

```diff
@@ -0,0 +1,138 @@
+    /// to work in environments such as docker on macOS, with network file systems such as NFS
+    /// or with mounted volumes in Kubernetes.
+    pub async fn poll(mut self) {
+        let mut interval = tokio::time::interval(TICK_INTERVAL);
+
+        loop {
+            interval.tick().await;
+
+            let Ok(fingerprint) = schema_fingerprint(&self.schema_path).await else {
+                continue;
+            };
+
+            let (fingerprint, schema) = match fingerprint {
+                Either::Left(modified) => (Either::Left(modified), None),
+                Either::Right((hash, schema)) => (Either::Right(hash), Some(schema)),
+            };
+
+            if fingerprint != self.schema_fingerprint {
+                self.schema_fingerprint = fingerprint;
+
+                let schema = match schema {
+                    Some(schema) => schema,
+                    None => {
+                        let Ok(schema) = tokio::fs::read_to_string(&self.schema_path).await else {
+                            tracing::warn!(
+                               "Could not load schema file in {}. The gateway will not use schema hot reload until file is available.",
+                               self.schema_path.to_str().unwrap_or_default()
+                           );
+
+                            continue;
+                        };
+
+                        schema
+                    }
+                };
+
+                tracing::info!("Detected a schema file update");
+
+                self.sender
+                    .send(GraphDefinition::Sdl(schema))
+                    .await
+                    .expect("channel must be up");
+            } else {
+                tracing::debug!("No schema file update detected");
+            }
+        }
+    }
+}
+
+async fn schema_fingerprint(schema_path: &Path) -> Result<Either<SystemTime, (u64, String)>, ()> {
+    let metadata = match tokio::fs::metadata(schema_path).await {
+        Ok(metadata) => metadata,
+        Err(_) => {
+            tracing::warn!(
+                "Could not load schema metadata from the filesystem. The gateway will not use schema hot reload until file is readable in {}.",
+                schema_path.to_str().unwrap_or_default(),
+            );
+
+            return Err(());
+        }
+    };
+
+    match metadata.modified() {
+        // If the file system supports modification date, we can use it to check for changes with less resources.
+        Ok(modified) => Ok(Either::Left(modified)),
+        // This is a fallback for file systems that do not support modification date. It loads and hashes the file
+        // contents to check for changes.
+        Err(_) => {
+            tracing::debug!("The file system with the schema file does not support modification date. The schema hot reload will use a bit more CPU by hashing the file contents.");
+
+            match tokio::fs::read_to_string(schema_path).await {
```

Comment:
Depending on how often that will happen (does it happen with configmaps?), there's an optimization to avoid excessive memory usage (with a tradeoff) that would be worth it here imo, to minimize the memory overhead: we don't need to read the whole file into memory, you can hash it chunk after chunk, and only if the hash changed, read the whole file into memory..

Reply:
Should not happen at all. I'm not sure how useful this part of the code even is, but this code is run only if the filesystem does not return metadata, especially of modification date is missing. So rare, but I want to write this optimization still.